### DISCUSSION TITLE
Unable to load Bool from CSV to SQLAlchemy

### DISCUSSION THREAD

**Discussion by delthia** [sqlalchemy/sqlalchemy#11738] [Usage Questions] [answered]
Hi,
I've been trying to load a csv file to an sqlalchemy db, I'm currently using with sqlite.

So the csv looks like this:
```
service_id,monday,tuesday,wednesday,thursday,friday,saturday,sunday,start_date,end_date
100010650,1,1,1,1,1,0,0,20240730,20250101
```
where the days of the week are booleans.

I'm trying to load it into this model:
```python
class Calendar(Base):
    __tablename__ = "calendar"

    service_id: Mapped[str] = mapped_column(primary_key=True)
    monday: Mapped[bool] = mapped_column(nullable=False)
    tuesday: Mapped[bool] = mapped_column(nullable=False)
    wednesday: Mapped[bool] = mapped_column(nullable=False)
    thursday: Mapped[bool] = mapped_column(nullable=False)
    friday: Mapped[bool] = mapped_column(nullable=False)
    saturday: Mapped[bool] = mapped_column(nullable=False)
    sunday: Mapped[bool] = mapped_column(nullable=False)
    start_date: Mapped[str] = mapped_column(nullable=False)
    end_date: Mapped[str] = mapped_column(nullable=False)
```

I'm reading the csv as follows:
```python
entries = []
# Read from file
with open('calendar.csv') as f:
    for row in csv.DictReader(f):
        entries.append(Calendar(**row))
# Save to database
with Session(self.engine) as session:
    session.add_all(entries)
    session.commit()
```

The problem is that `'0'` and `'1'` aren't converted into a boolean. I hopefully need a way for them to be converted into the model, as I'll load several csv into different objects in a loop, and wan't to avoid a long string of `if/else`s to detect each bool in each object type.
I tried casting the types by adding an `__init__` method to the object, but then I'd have to add all the optional parameters (I think), so it would become a bit tedious (there are more parameters than shown, those are the mandatory ones).

Can this be fixed with `TypeDecorator`? How should I use it?

Thanks

P.S.: Would loading it from sqlite by importing the csv be a problem? How would those types work with sqlalchemy if saved as a 0 or 1 directly with sqlite?

**Comment by CaselIT** [comment, marked as answer]
Hi,

Yes a type decorator can be used for this. Take a look at the recepies that do similar transformations https://docs.sqlalchemy.org/en/20/core/custom_types.html#typedecorator-recipes

**Reply by delthia** [reply]
Great, thanks a lot!
Would this work if I load the csv directly in sqlite and the work with those models in sqlalchemy?

**Reply by CaselIT** [reply]
Yes, you can change the types that is returned by the db with a type decorator

**Reply by delthia** [reply]
Ok, great. Thanks
